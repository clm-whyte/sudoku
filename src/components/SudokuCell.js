import React from "react";
import Button from "@material-ui/core/Button";
import { StylesProvider } from "@material-ui/core/styles";

export default function SudokuCell(props) {
  return (
    <StylesProvider injectFirst>
      <Button
        variant="outlined"
        aria-label={props.cellID}
        data-test={props.cellID}
        onClick={props.onClick}
        selected={props.selected}
        className={`sudokuCell
        ${props.row % 3 === 0 ? "boxBoundaryTop" : ""}
        ${props.row % 3 === 2 ? "boxBoundaryBottom" : ""}
        ${props.row === 8 ? "boxBoundaryBottom" : ""}
        ${props.col % 3 === 0 ? "boxBoundaryLeft" : ""}
        ${props.col % 3 === 2 ? "boxBoundaryRight" : ""}
        ${props.col === 8 ? "boxBoundaryRight" : ""}
        ${props.selected ? "selected" : ""}
        `}
      >
        {props.value}
      </Button>
    </StylesProvider>
  );
}
