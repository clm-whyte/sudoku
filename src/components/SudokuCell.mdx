import { Story, Preview } from "@storybook/addon-docs/blocks";
import SudokuCell from "./SudokuCell.js";

# Cell

In standard Sudoku, the grid is made up of 81 **cells** in a 9x9 square (9 rows and 9 columns).  
The **cells** are arranged into 3x3 sections called **boxes**, which are outlined in bold.

## Default

By default, a cell is has no value, is unselected and has a thin grey outline.  
The cells are **Material UI** Buttons which have been restyled using some Sassy CSS.

```css
.sudokuCell {
  max-width: 34px;
  max-height: $cellSize;
  min-width: 34px;
  min-height: $cellSize;
  border: $cellBorder;
  border-radius: 0;
  background-color: white;
  color: navy;
}
```

<Preview>
  <Story id="cell--empty" />
</Preview>

## Position

A cell's position in the grid determines its outline.  
If it's in the first row/column, it will have a thick top/left border respectively.  
Otherwise, every 3rd row/column is given a thick bottom/right border, based on the following lines of code:

```javascript
${props.col % 3 === 2 ? "boxBoundaryRight" : ""}
${props.row % 3 === 2 ? "boxBoundaryBottom" : ""}
```

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--outer-corner" />
  <Story id="cell--inner-corner" />
</Preview>

## Selection

By default a cell is unselected.  
An unselected cell has a white background and may or may not be filled with a value.

A selected cell is denoted by a pink background colour.

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--selected" />
</Preview>

## Cursor

There can be only one "cursor" Cell on the grid.  
The cursor shows the current position of the "active" cell. By default, this is at position `0, 0` and unselected.

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--cursor" />
  <Story id="cell--cursor-and-selected" />
</Preview>

## Value

The cell values default to a value of `null`.

<Preview>
  <Story id="cell--empty" />
  <Story id="cell--filled" />
</Preview>
