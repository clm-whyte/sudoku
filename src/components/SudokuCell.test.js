/// <reference types="jest" />
import React from "react";
import { shallow } from "enzyme";
import SudokuCell from "./SudokuCell";

describe("Cell borders", () => {
  const gridTopLeft = shallow(<SudokuCell row="0" col="0" />);
  const gridTopRight = shallow(<SudokuCell row="0" col="8" />);
  const gridBotLeft = shallow(<SudokuCell row="8" col="0" />);
  const gridBotRight = shallow(<SudokuCell row="8" col="8" />);

  it("should match the snapshot", () => {
    expect(gridTopLeft).toMatchSnapshot();
  });

  it("asserts that the cell in the top left should have a thick border on the top and left", () => {
    expect(gridTopLeft.html()).toContain("boxBoundaryLeft", "boxBoundaryTop");
  });

  it("asserts that the cell in the top right should have a thick border on the top and right", () => {
    expect(gridTopRight.html()).toContain("boxBoundaryRight", "boxBoundaryTop");
  });

  it("asserts that the cell in the bottom left should have a thick border on the bottom and left", () => {
    expect(gridBotLeft.html()).toContain(
      "boxBoundaryLeft",
      "boxBoundaryBottom"
    );
  });

  it("asserts that the cell in the top right should have a thick border on the top and right", () => {
    expect(gridBotRight.html()).toContain(
      "boxBoundaryRight",
      "boxBoundaryBottom"
    );
  });
});
