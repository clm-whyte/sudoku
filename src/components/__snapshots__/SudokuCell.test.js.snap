// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell borders should match the snapshot 1`] = `
<StylesProvider
  injectFirst={true}
>
  <WithStyles(ForwardRef(Button))
    className="sudokuCell
        
        
        boxBoundaryLeft
        boxBoundaryTop
        
        "
    variant="outlined"
  />
</StylesProvider>
`;
